lib/gooddata/lcm/actions/synchronize_users.rb

```diff
@@ -226,74 +230,71 @@ def call(params)
-                        puts "Project #{project.pid} of client #{client_id} will receive #{users.count} users"
-                        project.import_users(users,
-                                             domain: domain,
-                                             whitelists: whitelists,
-                                             ignore_failures: ignore_failures,
-                                             remove_users_from_project: remove_users_from_project,
-                                             do_not_touch_users_that_are_not_mentioned: do_not_touch_users_that_are_not_mentioned,
-                                             create_non_existing_user_groups: create_non_existing_user_groups)
+                        client_project = domain.clients(client_id).project
+                        fail "Client #{client_id} does not have project." unless client_project
+                        semaphore.synchronize { params.gdc_logger.info("Project #{client_project.pid} of client #{client_id} will receive #{users.count} users") }
+                        client_project.import_users(users,
+                                                    domain: domain,
+                                                    whitelists: whitelists,
+                                                    ignore_failures: ignore_failures,
+                                                    remove_users_from_project: remove_users_from_project,
+                                                    do_not_touch_users_that_are_not_mentioned: do_not_touch_users_that_are_not_mentioned)
                       end
                     when 'sync_domain_client_workspaces'
                       domain_clients = domain.clients
                       if params.segments_filter
                         segments_filter = params.segments_filter.map { |seg| "/gdc/domains/#{domain.name}/segments/#{seg}" }
                         domain_clients.select! { |c| segments_filter.include?(c.segment_uri) }
                       end
-                      working_client_ids = []
-                      res = []
-                      res += new_users.group_by { |u| u[:pid] }.flat_map do |client_id, users|
+                      working_client_ids = ThreadSafe::Array.new
+                      res = ThreadSafe::Array.new
+                      res += new_users.group_by { |u| u[:pid] }.flat_pmap do |client_id, users|
                         fail "Client id cannot be empty" if client_id.blank?
-                        c = domain.clients(client_id)
-                        if params.segments_filter && !segments_filter.include?(c.segment_uri)
-                          puts "Client #{client_id} is outside segments_filter #{params.segments_filter}"
+                        segment_client = domain_clients.find { |domain_client| domain_client.client_id.to_s == client_id.to_s }
+                        if params.segments_filter && !segments_filter.include?(segment_client.segment_uri)
```

Comment:
Identical code found in synchronize_user_filters.rb:174, synchronize_users.rb:256 (mass = 96)